crates/primitives/src/revm/env.rs

```diff
@@ -256,40 +277,74 @@ where
-            tx_env.gas_limit = *gas_limit;
-            tx_env.gas_price = U256::from(*max_fee_per_gas);
-            tx_env.gas_priority_fee = Some(U256::from(*max_priority_fee_per_gas));
-            tx_env.transact_to = match to {
-                TransactionKind::Call(to) => TransactTo::Call(*to),
+        Transaction::Eip4844(tx) => {
+            tx_env.gas_limit = tx.gas_limit;
+            tx_env.gas_price = U256::from(tx.max_fee_per_gas);
+            tx_env.gas_priority_fee = Some(U256::from(tx.max_priority_fee_per_gas));
+            tx_env.transact_to = match tx.to {
+                TransactionKind::Call(to) => TransactTo::Call(to),
                 TransactionKind::Create => TransactTo::create(),
             };
-            tx_env.value = (*value).into();
-            tx_env.data = input.clone();
-            tx_env.chain_id = Some(*chain_id);
-            tx_env.nonce = Some(*nonce);
-            tx_env.access_list = access_list
+            tx_env.value = tx.value.into();
+            tx_env.data = tx.input.clone();
+            tx_env.chain_id = Some(tx.chain_id);
+            tx_env.nonce = Some(tx.nonce);
+            tx_env.access_list = tx
+                .access_list
                 .0
                 .iter()
                 .map(|l| {
                     (l.address, l.storage_keys.iter().map(|k| U256::from_be_bytes(k.0)).collect())
                 })
                 .collect();
-            tx_env.blob_hashes = blob_versioned_hashes.clone();
-            tx_env.max_fee_per_blob_gas = Some(U256::from(*max_fee_per_blob_gas));
+            tx_env.blob_hashes = tx.blob_versioned_hashes.clone();
+            tx_env.max_fee_per_blob_gas = Some(U256::from(tx.max_fee_per_blob_gas));
+
+            #[cfg(feature = "optimism")]
+            fill_op_tx_env(tx_env, transaction, envelope);
+        }
+        #[cfg(feature = "optimism")]
+        Transaction::Deposit(tx) => {
+            tx_env.gas_limit = tx.gas_limit;
+            tx_env.gas_price = U256::ZERO;
+            tx_env.gas_priority_fee = None;
+            match tx.to {
+                TransactionKind::Call(to) => tx_env.transact_to = TransactTo::Call(to),
+                TransactionKind::Create => tx_env.transact_to = TransactTo::create(),
+            }
+            tx_env.value = tx.value.into();
+            tx_env.data = tx.input.clone();
+            tx_env.chain_id = None;
+            tx_env.nonce = None;
+
+            fill_op_tx_env(tx_env, transaction, envelope);
+        }
+    }
+}
+
+#[cfg(feature = "optimism")]
+#[inline(always)]
+fn fill_op_tx_env<T: AsRef<Transaction>>(tx_env: &mut TxEnv, transaction: T, envelope: Bytes) {
+    match transaction.as_ref() {
+        Transaction::Deposit(tx) => {
+            tx_env.optimism = OptimismFields {
+                source_hash: Some(tx.source_hash),
+                mint: tx.mint,
+                is_system_transaction: Some(tx.is_system_transaction),
+                enveloped_tx: Some(envelope),
```

Comment:
I think there's a really unfortunate requirement here to serialize the transaction to calculate its fee in revm and I wish we could calculate it w/o serializing (statically somehow?), otherwise this feels like it'll be a bottleneck if you try to do a lot of tps